File: app/src/main/java/com/tommihirvonen/exifnotes/Activities/MainActivity.java
package com.tommihirvonen.exifnotes.Activities;

import android.Manifest;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.drawable.ColorDrawable;
import android.location.LocationManager;
import android.os.Build;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.tommihirvonen.exifnotes.Dialogs.SimpleEula;
import com.tommihirvonen.exifnotes.Fragments.FramesFragment;
import com.tommihirvonen.exifnotes.Fragments.RollsFragment;
import com.tommihirvonen.exifnotes.R;
import com.tommihirvonen.exifnotes.Utilities.Utilities;

import java.io.File;

// Copyright 2015
// Tommi Hirvonen

/**
 * MainActivity is the first activity to be called when the app is launched.
 * It contains the RollsFragment and FramesFragment fragments.
 * The activity switches between these two fragments.
 */
public class MainActivity extends AppCompatActivity implements
        RollsFragment.OnRollSelectedListener,
        FramesFragment.OnHomeAsUpPressedListener {

    private final static int MY_MULTIPLE_PERMISSIONS_REQUEST = 1;
    boolean locationEnabled = false;

    /**
     * Create the layout and activate location services.
     *
     * @param savedInstanceState passed to super.onCreate to execute necessary code to properly
     *                           create the fragment
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        new SimpleEula(this).show();

        Utilities.setUiColor(this, false);
        if (getSupportActionBar() != null) getSupportActionBar().setTitle(
                "  " + getResources().getString(R.string.MainActivityTitle));

        LocationManager locationManager = (LocationManager) this.getSystemService(LOCATION_SERVICE);

        boolean permissionWriteExternalStorage = ActivityCompat.checkSelfPermission(
                this, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED;
        boolean permissionAccessCoarseLocation = ActivityCompat.checkSelfPermission(
                this, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED;
        boolean permissionAccessFineLocation = ActivityCompat.checkSelfPermission(
                this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED;

        //Check if the app has all necessary permissions
        if (!permissionWriteExternalStorage || !permissionAccessCoarseLocation || !permissionAccessFineLocation) {
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE,
                                Manifest.permission.ACCESS_COARSE_LOCATION,
                                Manifest.permission.ACCESS_FINE_LOCATION},
                    MY_MULTIPLE_PERMISSIONS_REQUEST);
        } else {
            locationEnabled = true;
        }



        // getting GPS status
        boolean isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
        if (!isGPSEnabled) showSettingsAlert();


        // Check that the activity is using the layout version with
        // the fragment_container FrameLayout
        if (findViewById(R.id.fragment_container) != null) {

            // However, if we're being restored from a previous state,
            // then we don't need to do anything and should return or else
            // we could end up with overlapping fragments.
            if (savedInstanceState != null) {
                return;
            }

            // Create a new Fragment to be placed in the activity layout
            RollsFragment firstFragment = new RollsFragment();

            // Add the fragment to the 'fragment_container' FrameLayout
            getFragmentManager().beginTransaction()
                    .add(R.id.fragment_container, firstFragment, RollsFragment.ROLLS_FRAGMENT_TAG).commit();
        }
    }

    /**
     * Inflate the main menu.
     *
     * @param menu the menu to be inflated
     * @return super class to execute code for the menu to work properly.
     */
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);

        return super.onCreateOptionsMenu(menu);
    }

    /**
     * Called when the user selects a menu item.
     *
     * @param item The menu item that was selected.
     * @return super so that other handlers can handle the selection
     */
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        return super.onOptionsItemSelected(item);
    }

    /**
     * When resuming we have to change the UI colours again.
     */
    @Override
    public void onResume(){
        super.onResume();

        int primaryColor = Utilities.getPrimaryUiColor(getBaseContext());
        int secondaryColor = Utilities.getSecondaryUiColor(getBaseContext());
        Utilities.setSupportActionBarColor(this, primaryColor);
        Utilities.setStatusBarColor(this, secondaryColor);
    }

    /**
     * This function is called when the user presses the back button.
     * If the FramesFragment is in front then pop the stack back by one
     * and bring the RollsFragment to front.
     * Also bring the shadow under the action bar to front.
     */
    @Override
    public void onBackPressed() {

        int count = getFragmentManager().getBackStackEntryCount();

        if (count == 0) {
            super.onBackPressed();
            //additional code
        } else {
            getFragmentManager().popBackStack();
            View shadow = findViewById(R.id.shadow);
            shadow.bringToFront();
        }

    }

    /**
     * This function is called when MainActivity is started, in other words when the
     * application is started. All the files created in FramesFragment.setShareIntentExportRoll
     * in the application's external storage directory are deleted. This way we can keep
     * the number of files stored to a minimum.
     */
    @Override
    public void onStart(){
        //Delete all the files created in FramesFragment.setShareIntentExportRoll
        File externalStorageDir = getExternalFilesDir(null);
        Utilities.purgeDirectory(externalStorageDir);
        super.onStart();
    }

    /**
     * This function is called when the user presses a roll in the RollsFragment.
     * Create a new FramesFragment, pass the roll id and locationEnabled to it and bring it to the
     * front of the stack.
     * Also bring the shadow under the action bar to front.
     *
     * @param rollId The id of the roll that was pressed.
     */
    @Override
    public void onRollSelected(long rollId){
        FramesFragment newFragment = new FramesFragment();
        Bundle args = new Bundle();
        args.putLong("ROLL_ID", rollId);
        args.putBoolean("LOCATION_ENABLED", locationEnabled);
        newFragment.setArguments(args);
        getFragmentManager().beginTransaction()
                .setCustomAnimations(
                        R.animator.slide_left,
                        R.animator.slide_right,
                        R.animator.slide_left,
                        R.animator.slide_right)
                .replace(R.id.fragment_container, newFragment, FramesFragment.FRAMES_FRAGMENT_TAG)
                .addToBackStack(null).commit();
        View shadow = findViewById(R.id.shadow);
        shadow.bringToFront();
    }


    /**
     * When the user presses the home as up button in the action bar this function is called.
     * Pop the FramesFragment off the stack.
     */
    public void onHomeAsUpPressed(){
        getFragmentManager().popBackStack();
    }


    /**
     * This function is called if GPS is not enabled.
     * Prompt the user to jump to settings to enable GPS.
     */
    public void showSettingsAlert(){
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(this);

        // Setting Dialog Title
        alertDialog.setTitle(R.string.GPSSettings);

        // Setting Dialog Message
        alertDialog.setMessage(R.string.GPSNotEnabled);

        // On pressing Settings button
        alertDialog.setPositiveButton(R.string.Settings, new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog,int which) {
                Intent intent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
                startActivity(intent);
            }
        });

        // on pressing cancel button
        alertDialog.setNegativeButton(R.string.Cancel, new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        });

        // Showing Alert Message
        alertDialog.show();
    }

    /**
     * When the user is asked for permission to use location services, handle the request result.
     *
     * @param requestCode The code of the request that was made.
     * @param permissions not used
     * @param grantResults An array which is empty if the request was cancelled.
     */
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults) {
        switch (requestCode) {

            case MY_MULTIPLE_PERMISSIONS_REQUEST:

                // If request is cancelled, the result arrays are empty. Thus we check
                // the length of grantResults first.

                //Check location permissions
                if (grantResults.length > 0 && grantResults[1] == PackageManager.PERMISSION_GRANTED
                        && grantResults[2] == PackageManager.PERMISSION_GRANTED) {
                    locationEnabled = true;
                }

                //Check write permissions
                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_DENIED) {
                    //In case write permission was denied, inform the user.
                    Toast.makeText(this, R.string.NoWritePermission, Toast.LENGTH_LONG).show();
                }

                break;
        }
    }
}

File: app/src/main/java/com/tommihirvonen/exifnotes/activities/MainActivity.kt
/*
 * Exif Notes
 * Copyright (C) 2022  Tommi Hirvonen
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tommihirvonen.exifnotes.activities

import android.Manifest
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.location.LocationManager
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.app.ActivityCompat
import androidx.preference.PreferenceManager
import com.google.android.libraries.places.api.Places
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.tommihirvonen.exifnotes.R
import com.tommihirvonen.exifnotes.data.Database
import com.tommihirvonen.exifnotes.databinding.ActivityMainBinding
import com.tommihirvonen.exifnotes.dialogs.TermsOfUseDialog
import com.tommihirvonen.exifnotes.preferences.PreferenceConstants
import com.tommihirvonen.exifnotes.utilities.ComplementaryPicturesManager
import com.tommihirvonen.exifnotes.utilities.purgeDirectory
import com.tommihirvonen.exifnotes.utilities.snackbar
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

/**
 * MainActivity is the first activity to be called when the app is launched.
 * It contains the RollsFragment and FramesFragment fragments.
 * The activity switches between these two fragments.
 */
@AndroidEntryPoint
class MainActivity : AppCompatActivity() {

    @Inject
    lateinit var database: Database

    companion object {
        /**
         * Tag for permission request
         */
        private const val MY_MULTIPLE_PERMISSIONS_REQUEST = 1
    }

    private val binding by lazy { ActivityMainBinding.inflate(layoutInflater) }

    override fun onCreate(savedInstanceState: Bundle?) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(this)

        // If the app theme was set in the app's preferences, override the night mode setting.
        when (prefs.getString(PreferenceConstants.KEY_APP_THEME, "DEFAULT") ?: "DEFAULT") {
            "LIGHT" -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
            "DARK" -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        }

        super.onCreate(savedInstanceState)

        // Delete all complementary pictures, which are not linked to any frame.
        // Do this each time the app is launched to keep the storage consumption to a minimum.
        // If savedInstanceState is not null, then the activity is being recreated. In this case,
        // don't delete pictures.
        if (savedInstanceState == null) {
            ComplementaryPicturesManager.deleteUnusedPictures(this, database)
        }

        setContentView(binding.root)

        if (savedInstanceState == null) {
            TermsOfUseDialog(this).show()

            // Check that the application has write permission to the phone's external storage
            // and access to location services.
            val permissions = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                arrayOf(
                    Manifest.permission.ACCESS_COARSE_LOCATION,
                    Manifest.permission.ACCESS_FINE_LOCATION,
                    Manifest.permission.POST_NOTIFICATIONS
                )
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                arrayOf(
                    Manifest.permission.ACCESS_COARSE_LOCATION,
                    Manifest.permission.ACCESS_FINE_LOCATION
                )
            } else {
                arrayOf(
                    Manifest.permission.WRITE_EXTERNAL_STORAGE,
                    Manifest.permission.ACCESS_COARSE_LOCATION,
                    Manifest.permission.ACCESS_FINE_LOCATION
                )
            }
            val permissionGrants = permissions.map {
                ActivityCompat.checkSelfPermission(this, it) == PackageManager.PERMISSION_GRANTED
            }
            if (permissionGrants.contains(false)) {
                ActivityCompat.requestPermissions(this, permissions, MY_MULTIPLE_PERMISSIONS_REQUEST)
            }

            // Get from DefaultSharedPreferences whether the user has enabled
            // location updates in the app's settings.
            val requestingLocationUpdates = prefs.getBoolean(PreferenceConstants.KEY_GPS_UPDATE, true)

            // Getting GPS status
            val locationManager = this.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            val isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)

            // Show a dialog to go to settings only if GPS is not enabled in system settings
            // but location updates are enabled in the app's settings.
            if (!isGPSEnabled && requestingLocationUpdates) showSettingsAlert()
        }

        if (!Places.isInitialized()) {
            val apiKey = resources.getString(R.string.google_maps_key)
            Places.initialize(applicationContext, apiKey)
        }
    }

    override fun onStart() {
        // This method is called when MainActivity is started, in other words when the
        // application is started. All the files created in FramesFragment.setShareIntentExportRoll
        // in the application's external storage directory are deleted. This way we can keep
        // the number of files stored to a minimum.

        //Delete all the files created in FramesFragment.setShareIntentExportRoll
        val externalStorageDir = getExternalFilesDir(null)
        externalStorageDir?.purgeDirectory()
        externalCacheDir?.purgeDirectory()
        super.onStart()
    }

    /**
     * This method is called if GPS is not enabled.
     * Prompt the user to jump to device settings to enable GPS.
     */
    private fun showSettingsAlert() = MaterialAlertDialogBuilder(this)
        .setTitle(R.string.GPSSettings)
        .setMessage(R.string.GPSNotEnabled)
        // Navigate to the device's settings.
        .setPositiveButton(R.string.GoToSettings) { _: DialogInterface?, _: Int ->
            val intent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
            startActivity(intent)
        }
        // Disable location tracking in the application's settings.
        .setNegativeButton(R.string.DisableInApp) { dialogInterface: DialogInterface, _: Int ->
            val prefs = PreferenceManager.getDefaultSharedPreferences(baseContext)
            val prefsEditor = prefs.edit()
            prefsEditor.putBoolean(PreferenceConstants.KEY_GPS_UPDATE, false)
            prefsEditor.apply()
            dialogInterface.dismiss()
        }
        .setNeutralButton(R.string.Cancel) { dialog: DialogInterface, _: Int -> dialog.cancel() }
        .show()

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        if (requestCode == MY_MULTIPLE_PERMISSIONS_REQUEST) {
            // If request is cancelled, the result arrays are empty. Thus we check the length of grantResults first.
            // Check write permissions
            val writePermissionIndex = permissions.indexOf(Manifest.permission.WRITE_EXTERNAL_STORAGE)
            if (grantResults.getOrNull(writePermissionIndex) == PackageManager.PERMISSION_DENIED) {
                //In case write permission was denied, inform the user.
                binding.root.snackbar(R.string.NoWritePermission)
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

}